Clarify that the license is GPLv2+

sledgehammer999 opened the pull request:
Regarding this license clarification there are 3 commits of interest
(commits A, B, C). Before commit A the COPYING file contained only the
text of the GPLv2 license, while all source files had a license block
at the top saying that they are under the terms of "GPLv2 or later". With
commit A there was a temporary change to GPLv3. The COPYING file contained
only the text of the GPLv3 license, while all source files had a license
block at the top saying that they are under the terms of "GPLv3 or later".
Then with commit B the COPYING file and the license block of the source
files was reverted to their state before commit A. Afterwards, with
commit C a license summary(or clarification) block was put at the top of
the COPYING file. This block indicated that the license was GPLv2 without
having the "or later" clause and it also included the OpenSSL exception.
However, the license block of each source file continued to contain the
"or later" clause which was not removed. The same license block continues
to exist in all current source files. Thus it is concluded that the ommision
of the "or later" clause with commit C in the COPYING file was accidental.
~OR ALTERNATIVELY (OR IN ADDITION)
Maybe with commits B and C it was decided that the license of the
project(qBittorrent) as a whole/bundle was indeed GPLv2 only, while the
individual source files were to remain as GPLv2+ so any 3rd party could
use them under the more relaxed GPLv2+ terms. It should be noted that new
code and source files also have had the GPLv2+ terms. So from this point
on we exercise the same rights the 3rd parties have and choose GPLv2+ terms
for the project too.~
OR ALTERNATIVELY (OR IN ADDITION)
At the time commit C was made Christophe Dumez was not the sole contributor.
There is no record that the other contributors agreed with the supposed
GPLv2 only change or that there was a Contributor License Agreement,
transfering their rights to him. Thus making his license change decision
invalid/void/illegal.

Commit A: 54f9375b325cc03f8d5da9abbf6c4e25e8ee698f
Commit B: 8df61db644a7fcb19170d60da3b7b5297215439c
Commit C: 9835af49627e65d314668387b6ec9d029707c186

sledgehammer999 commented:
I created the following program to check if source files had the "or later" clause before commit A and in current HEAD. I post the code here so you can verify I didn't make a mistake. With it I confirmed that the "or later" clause existed then and exists now in all source files. There are some exceptions but those files have compatible licenses (taken from 3rd parties).

```c++
#include <iostream>

#include <QCoreApplication>
#include <QDir>
#include <QFile>
#include <QString>
#include <QStringList>
#include <QTextStream>

namespace {
    const QString basePath = "G:/QBITTORRENT/qBittorrent/src";
    const QStringList extenstions {"*.c", "*.cpp", "*.cxx", "*.mm", "*.h", "*.hpp"};
}

bool checkGPLplus(QTextStream &stream) {
    QString line = stream.readLine();
    if (line.startsWith(" * modify it under the terms of the GNU General Public License")) {
        line = stream.readLine();
        if (line.startsWith(" * as published by the Free Software Foundation; either version 2")) {
            line = stream.readLine();
            if (line.startsWith(" * of the License, or (at your option) any later version.")) {
                return true;
            }
        }
    }

    return false;
}

bool containsLaterClause(const QString &filename) {
    QFile file(filename);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return false;
    QTextStream stream(&file);

    for (int i = 0; ((i < 13) && !stream.atEnd()); ++i) {
        QString line = stream.readLine();
        if (line.startsWith(" * This program is free software; you can redistribute it and/or"))
            return checkGPLplus(stream);
    }

    return false;
}

QStringList listFiles(const QString &dirName) {
    const QDir dir(dirName);
    const QStringList files = dir.entryList(extenstions, QDir::Files);
    const QStringList folders = dir.entryList(QDir::Dirs | QDir::NoDotAndDotDot);

    QStringList absoluteNames;
    for (const auto &file : files)
        absoluteNames.append(dir.absoluteFilePath(file));

    for (const auto &folder : folders)
        absoluteNames.append(listFiles(dir.absoluteFilePath(folder)));

    return absoluteNames;
}

int main()
{
    const QStringList files = listFiles(basePath);
    std::cout << "Found source files: " << files.size() << std::endl;

    QStringList failed;
    for (const auto &file : files) {
        if (!containsLaterClause(file))
            failed.append(file);
    }

    std::cout << "Failed: " << failed.size() << std::endl;
    for (const auto &fail : failed)
        std::cout << fail.toLocal8Bit().constData() << std::endl;

    return 0;
}

```

Chocobo1 commented on AUTHORS:
```diff
@@ -17,9 +17,13 @@ Contributors:
+  copyright: Trolltech ASA <info@trolltech.com>
+  license: custom
 
 * files src/ico.cpp src/ico.h
```
We don't have this file entry now.

Chocobo1 commented on AUTHORS:
```diff
@@ -17,9 +17,13 @@ Contributors:
-  license: LGPL
+  license: mixed
+
+* files src/gui/lineedit.*
```
IIRC this file has received many changes, but I guess that doesn't mean we could stop using their license?

sledgehammer999 replied:
If the contributors didn't care to restrict their contributions they fall under that license.

glassez replied:
Of course, I didn't fully understand everything, but reading [GPL FAQ](https://www.gnu.org/licenses/gpl-faq.en.html), I see that the license applies to the whole program.
We may obtain, use, modify, and distribute the program under its license. We may combine several (possibly modified) programs into one and distribute it under a license that meets the conditions of all the combined source programs.
If we borrow some individual source files from a program, we are actually combining a modified version of that program with our program. So the entire resulting product must be distributed under a single license (or several license each of) that applies to the entire product (to all of its files).

Chocobo1 commented:
>Maybe with commits B and C it was decided that the license of the project(qBittorrent) as a whole/bundle was indeed GPLv2 only, while the individual source files were to remain as GPLv2+ so any 3rd party could use them under the more relaxed GPLv2+ terms. It should be noted that new code and source files also have had the GPLv2+ terms. 

The only license that is fully compatible with GPL2 is GPL3 and AFAIK GPL3 is even more restricted/well-defined than its predecessor. So IMO this alternate possibility doesn't make sense.

glassez commented:
> The only license that is fully compatible with GPL2 is GPL3 and AFAIK GPL3 is even more restricted/well-defined than its predecessor. So IMO this alternate possibility doesn't make sense.

What do you mean/suggest exactly?

Chocobo1 commented:
>What do you mean/suggest exactly?

Just saying my opinion on that paragraph.
As for my preference I would hope to see this project under GPLv2+ license.

glassez commented:
> As for my preference I would hope to see this project under GPLv2+ license.

:+1:
I still don't understand what you meant in the next sentence:
>IMO this imaginary possibility doesn't make sense.

sledgehammer999 commented:
>>Maybe with commits B and C it was decided that the license of the project(qBittorrent) as a whole/bundle was indeed GPLv2 only, while the individual source files were to remain as GPLv2+ so any 3rd party could use them under the more relaxed GPLv2+ terms.
>
>The only license that is fully compatible with GPL2 is GPL3 and AFAIK GPL3 is even more restricted/well-defined than its predecessor. So IMO this imaginary possibility doesn't make sense.

Maybe I need to rephrase that part to make it clearer? The individual source files are GPLv2+, while the project as a whole is GPLv2 only. Third parties could copy one of the individual files under GPLv2+. So why can't we do the same?

glassez commented:
>The individual source files are GPLv2+, while the project as a whole is GPLv2 only. 

IMO, it's nonsense.

sledgehammer999 commented:
>IMO, it's nonsense.

Then ommit that alternative interpretation from the rationale?

glassez commented:
> > IMO, it's nonsense.
> 
> Then omit that alternative interpretation from the rationale?

Yes, IMO.
The main thing is that the headers containing `GPLv2 and above` clause should be added by the copyright holder.

glassez commented:
To clarify my point again... I treat the difference between COPYING and source files as an error/typo. Nothing prevents the copyright owner from changing the product license. Anyone who has received a copy of the product under the terms of GPLv2+ is also free to distribute it under the terms of GPLv2 (only). But then he had to change the conditions in all the files, not just in COPYING (which led to discrepancies). No wonder the FSF guys insist that conditions are contained in every source file: https://www.gnu.org/licenses/gpl-faq.en.html#NoticeInSourceFile. Following their reasoning, we may well prefer exactly the conditions that are specified in the source files.

sledgehammer999 commented:
I changed the commit rationale and removed the outdated entry from AUTHORS.

>No wonder the FSF guys insist that conditions are contained in every source file: https://www.gnu.org/licenses/gpl-faq.en.html#NoticeInSourceFile. Following their reasoning, we may well prefer exactly the conditions that are specified in the source files.

Yeah, I agree with what they're saying. I pretty much had come to the same conclusion. Each file should have it's own noticed for the reasons stated in that link.

Chocobo1 commented:
@sledgehammer999 
Note that we have a src/gui/gpl.html which should be updated as well.

FranciscoPombal commented:
Thanks @sledgehammer999 :+1:

Please see my comment in the related issue at https://github.com/qbittorrent/qBittorrent/issues/14571#issuecomment-804015552.

Would you prefer to continue all further discussion here rather than on that Issue post? If so, let me know and I'll close that issue. Not matter what happens, it's safe to say some form of this PR will get merged anyway, so that issue can be considered addressed even before this is officially merged.

FranciscoPombal commented:
https://github.com/qbittorrent/qBittorrent/tree/969a02b93e704a7cc7157780194b4ad267362bec was the first-ever commit to GitHub. Is there any public history before that (if so, where)? The corresponding commit message hints at a previous existence in an SVN repo. If there is previous public history, we may need to also take that into account (to figure out if the `COPYING` appeared before the headers in the files or not, and if the headers in the files changed over time).

glassez commented:
>to figure out if the COPYING appeared before the headers in the files or not, and if the headers in the files changed over time

I think it doesn't matter. At that time, when COPYING contained only the text of the license, it did not matter (according to the guys from the FSF).

FranciscoPombal commented:
> I think it doesn't matter. At that time, when COPYING contained only the text of the license, it did not matter (according to the guys from the FSF).

Sorry, this is a bit unclear to me. What do you mean (specifically the part about "according to the guys from the FSF").

glassez commented:
> specifically the part about "according to the guys from the FSF"

I refer to https://www.gnu.org/licenses/gpl-faq.en.html



> Sorry, this is a bit unclear to me. What do you mean

They say that you must provide copy of license with your program (according to GPL terms) but the presence of the license text in some file does not mean that the program is distributed under this license. So we have to put the appropriate statement:
>This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version. [...]

FranciscoPombal commented:
@glassez @sledgehammer999 

> I refer to https://www.gnu.org/licenses/gpl-faq.en.html
>
> They say that you must provide copy of license with your program (according to GPL terms) but the presence of the license text in some file does not mean that the program is distributed under this license. So we have to put the appropriate statement:

Alright. For anyone else interested, here's links to the relevant points:

https://www.gnu.org/licenses/gpl-faq.en.html#WhyMustIInclude
https://www.gnu.org/licenses/gpl-faq.en.html#LicenseCopyOnly.

Looks like it's not obligatory to include the verbatim text like I said, the only important part is the header, and as I've stated, there is no need to include it in every file.

They even state:

> This statement should be in each source file. A clear statement in the program's README file is legally sufficient _as long as that accompanies the code_, but it is easy for them to get separated.

I still argue that "it is easy for them to get separated" is not a realistic concern. But I don't mind including the full verbatim text of the license too much tbh.

In conclusion: we should have a LICENSE file including both the header (this is the most important and only vital part), and the verbatim copy of the GPLv2. Then we can delete the header from all source files.

sledgehammer999 commented:
I made the suggested changes (@Chocobo1 ) and added the webui file too.

Regarding the headers and license file: Well I am not big on changing things for the sake of change. Both `LICENSE` and `COPYING` are considered [valid](https://stackoverflow.com/questions/5678462/should-i-provide-a-license-txt-or-copying-txt-file-in-my-project) ways to indicate the license. So there's no need to switch to `LICENSE`.
Also I don't want to remove the headers from each source file.

If there aren't any other issues please review and approve this PR.

FranciscoPombal commented on COPYING:
```diff
@@ -10,6 +11,8 @@ modify file(s), you may extend this exception to your version of the file(s),
 but you are not obligated to do so. If you do not wish to do so, delete this
 exception statement from your version.
 
+Some files have their own licesnes. Take a look at the AUTHORS file.
```
Typo, `licesnes` -> `licenses`.

FranciscoPombal commented:
@sledgehammer999 

> I made the suggested changes (@Chocobo1 ) and added the webui file too.
> 
> Regarding the headers and license file: Well I am not big on changing things for the sake of change. Both `LICENSE` and `COPYING` are considered [valid](https://stackoverflow.com/questions/5678462/should-i-provide-a-license-txt-or-copying-txt-file-in-my-project) ways to indicate the license. So there's no need to switch to `LICENSE`.

Yeah but `LICENSE` is more obvious nowadays. `COPYING` is only the standard in the GNU project. But whatever, I agree that the naming of the file is the least of our worries.

> Also I don't want to remove the headers from each source file.
> 
> If there aren't any other issues please review and approve this PR.

Fine, I suppose the removal of the headers from each source file can be proposed separately. Obviously, such removal excludes the files that have different licenses (i.e. files that contain code from other projects), as there still has to be some indication that these files have a different license.

I just posted my review, just found a small typo.

glassez commented on COPYING:
```diff
@@ -10,6 +11,8 @@ modify file(s), you may extend this exception to your version of the file(s),
 but you are not obligated to do so. If you do not wish to do so, delete this
 exception statement from your version.
 
+Some files have their own licenses. Take a look at the AUTHORS file.
```
This statement (and all related stuff) contradicts the explanations from GPL FAQ in my eyes.
Entire Program (all its sources including borrowed from another Programs) should be distributed under the same license. If some Program license disallow you to distribute it (or its part) under GPL (i.e. Program has GPL incompatible license) you cannot include it (or it's part) into Program that is distributed under GPL.

sledgehammer999 replied:
>If some Program license disallow you to distribute it (or its part) under GPL (i.e. Program has GPL incompatible license) you cannot include it (or it's part) into Program that is distributed under GPL.

Yes. But I wasn't refering to that. I was refering to taking eg a MIT licensed source and using it. We can use it in GPL. But also that file has remained in MIT license, since we haven't explicitly relicensed it (and I don't know if relicensing was possible).

>This statement (and all related stuff) contradicts the explanations from GPL FAQ in my eyes.

Is there a direct FAQ answer or is it from several ones?

sledgehammer999 replied:
Also `AUTHORS` has some non-source files.

glassez replied:
https://www.gnu.org/licenses/gpl-faq.en.html#WhatDoesCompatMean

sledgehammer999 replied:
When we distribute qBittorrent we do that as a combination of code and that combination is released under the same GNU GPL version. However the 3rd party files, when taken individually out of the project, can **also** be used under their original license. If they are modified by us we could make them GPLv2+ only, but it would need to be an explicit action. I don't think any contributor has done that, thus those modifications can be used under the original license too.
Well, that's my interpretation of all that.

sledgehammer999 replied:
Maybe change the last sentence to:
```
Some 3rd party source files are under different licenses. Also non-source files are under their own licenses. See the AUTHORS file for details. It also contains some copyright info about the aforementioned files.
```

?

glassez replied:
>However the 3rd party files, when taken individually out of the project, can also be used under their original license.

Yes, if they are taken as part of their original program/project.
> If they are modified by us we could make them GPLv2+ only, but it would need to be an explicit action. 

Don't understand what explicit action you mean.

Let's look at an example of a lifecycle segment of arbitrary qBittorrent.
1. Someone got a qBittorrent v1 under GPL and "another app" under a compatible license, took the `file1` file from "another app", added it to qBittorrent, and then distributed a modified version of qBittorrent (v2) under the GPL. Now there are, as it were, two versions of the same file `file1`, distributed under different license (in different projects).
2. Someone got qBittorrent v2 under the GPL, modified it, including by modifying the `file1` file, and then distributed the modified version of qBittorrent (v3) under the GPL. Now a modified version of the `file1` file exists only as part of a program (qBittorrent v3) licensed under the GPL.

In any case, it seems to me incorrect to say about some qBittorrent files that they have their own license (unless they belong to those files that, in principle, do not fall under the GPL).

sledgehammer999 replied:
It's like the Qt where it has some modules under GPLv3 only. The GPLv2+ Qt modules continue to be under the GPLv2+ license if they are used separately from the GPLv3 modules.

>Don't understand what explicit action you mean.

Let's say I modified a 3rd party file that I copied from a compatible project. If I  didn't explicitly state* that my modifications to this file are under GPLv2+, then they are under the license that is mentioned at the top of the file.

*I could make such statement by adding a modification to the license block of the 3rd party file.

But the details of this issue don't matter much to us. Any person wishing to copy said file should do their due diligence and decide under which license the modified file is and what to do with it.
So in order for this to go forward how about I change the offending sentene to simply:
```See also the AUTHORS file```?

glassez replied:
>So in order for this to go forward how about I change the offending sentene to simply:
```See also the AUTHORS file```?

Ok, let's go with it.

sledgehammer999 replied:
done

glassez replied:
PR is approved to go further. But I will still add something else, as food for thought.

When someone makes public contribution in GPL project, it is obvious that they expect their work to be distributed under the terms of the GPL. For example, someone may not want their work to be used in closed source programs, etc. Putting a declaration that some files have a different license, we introduce confusion. Ideally, we should only mention that such and such files came from such and such projects that have such and such licenses, and not create the false impression that we distribute them under their own licenses.
>It's like the Qt where it has some modules under GPLv3 only. The GPLv2+ Qt modules continue to be under the GPLv2+ license if they are used separately from the GPLv3 modules.

Qt explicitly distributes some modules separately as an option. qBittorrent doesn't distribute any parts separately.
>Let's say I modified a 3rd party file that I copied from a compatible project. If I didn't explicitly state that my modifications to this file are under GPLv2+, then they are under the license that is mentioned at the top of the file.

Seems you still don't understand. When you contribute to qBittorrent, you modify qBittorrent's files under the terms of qBittorrent's license (even if some of these files were ever borrowed from a third-party project that has a different license).

xavier2k6 commented:
Not the greatest of sources I know, but Wikipedia declares qBittorrent as GPLv2+ ref.: https://en.wikipedia.org/wiki/QBittorrent
